File: scala/src/main/scala/algorithms/MergeSort.scala
package algorithms

import scala.reflect.ClassTag

/**
 * Merge sort - stable, divide-and-conquer sorting algorithm.
 * Time: O(n log n)
 * Space: O(n)
 */
object MergeSort:
  def sort[T: ClassTag](arr: Array[T])(using ord: Ordering[T]): Array[T] =
    if arr.length <= 1 then arr
    else
      mergeSort(arr, 0, arr.length - 1)
      arr

  private def mergeSort[T: ClassTag](arr: Array[T], left: Int, right: Int)(using ord: Ordering[T]): Unit =
    if left < right then
      val mid = left + (right - left) / 2
      mergeSort(arr, left, mid)
      mergeSort(arr, mid + 1, right)
      merge(arr, left, mid, right)

  private def merge[T: ClassTag](arr: Array[T], left: Int, mid: Int, right: Int)(using ord: Ordering[T]): Unit =
    val leftArr = arr.slice(left, mid + 1)
    val rightArr = arr.slice(mid + 1, right + 1)

    var i = 0
    var j = 0
    var k = left

    while i < leftArr.length && j < rightArr.length do
      if ord.lteq(leftArr(i), rightArr(j)) then
        arr(k) = leftArr(i)
        i += 1
      else
        arr(k) = rightArr(j)
        j += 1
      k += 1

    while i < leftArr.length do
      arr(k) = leftArr(i)
      i += 1
      k += 1

    while j < rightArr.length do
      arr(k) = rightArr(j)
      j += 1
      k += 1

  def sorted[T: ClassTag](arr: Array[T])(using ord: Ordering[T]): Array[T] =
    val copy = arr.clone()
    sort(copy)

File: go/algorithms/merge_sort.go
package algorithms

import "cmp"

// MergeSort sorts a slice using merge sort.
// Time: O(n log n) all cases
// Space: O(n)
func MergeSort[T cmp.Ordered](arr []T) {
	if len(arr) < 2 {
		return
	}
	mergeSort(arr, 0, len(arr)-1)
}

func mergeSort[T cmp.Ordered](arr []T, left, right int) {
	if left >= right {
		return
	}

	mid := left + (right-left)/2
	mergeSort(arr, left, mid)
	mergeSort(arr, mid+1, right)
	merge(arr, left, mid, right)
}

func merge[T cmp.Ordered](arr []T, left, mid, right int) {
	leftArr := make([]T, mid-left+1)
	rightArr := make([]T, right-mid)

	copy(leftArr, arr[left:mid+1])
	copy(rightArr, arr[mid+1:right+1])

	i, j, k := 0, 0, left

	for i < len(leftArr) && j < len(rightArr) {
		if leftArr[i] <= rightArr[j] {
			arr[k] = leftArr[i]
			i++
		} else {
			arr[k] = rightArr[j]
			j++
		}
		k++
	}

	for i < len(leftArr) {
		arr[k] = leftArr[i]
		i++
		k++
	}

	for j < len(rightArr) {
		arr[k] = rightArr[j]
		j++
		k++
	}
}

// MergeSortFunc sorts a slice using a custom comparison function.
func MergeSortFunc[T any](arr []T, less func(a, b T) bool) {
	if len(arr) < 2 {
		return
	}
	mergeSortFunc(arr, 0, len(arr)-1, less)
}

func mergeSortFunc[T any](arr []T, left, right int, less func(a, b T) bool) {
	if left >= right {
		return
	}

	mid := left + (right-left)/2
	mergeSortFunc(arr, left, mid, less)
	mergeSortFunc(arr, mid+1, right, less)
	mergeFunc(arr, left, mid, right, less)
}

func mergeFunc[T any](arr []T, left, mid, right int, less func(a, b T) bool) {
	leftArr := make([]T, mid-left+1)
	rightArr := make([]T, right-mid)

	copy(leftArr, arr[left:mid+1])
	copy(rightArr, arr[mid+1:right+1])

	i, j, k := 0, 0, left

	for i < len(leftArr) && j < len(rightArr) {
		if !less(rightArr[j], leftArr[i]) {
			arr[k] = leftArr[i]
			i++
		} else {
			arr[k] = rightArr[j]
			j++
		}
		k++
	}

	for i < len(leftArr) {
		arr[k] = leftArr[i]
		i++
		k++
	}

	for j < len(rightArr) {
		arr[k] = rightArr[j]
		j++
		k++
	}
}
